Add Environment Variable override of telemetry

lucyllewy opened the pull request:
## PR Summary

Inspired by dotnet cli, and allows for Telemetry optout on packaing systems such as AppImage and Snapcraft where the filesystem is immutable.

## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Issue filed - Issue link: https://github.com/PowerShell/PowerShell-Docs/issues/2088
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [NA] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
    - [NA] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

Commit:
Add Environment Variable override of telemetry

Inspired by dotnet cli, and allows for Telemetry optout on packaing systems such as AppImage and Snapcraft where the filesystem is immutable.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=6063) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=6063) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/PowerShell/PowerShell?pullRequest=6063) it.</sub>

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -21,6 +21,7 @@ internal static class ApplicationInsightsTelemetry
         // The name of the file by when present in $PSHOME will enable telemetry.
         // If this file is not present, no telemetry will be sent.
         private const string TelemetrySemaphoreFilename = "DELETE_ME_TO_DISABLE_CONSOLEHOST_TELEMETRY";
+        private const string TelemetryOptout = "POWERSHELL_TELEMETRY_OPTOUT";
```
Need rename the file in root repo folder.

SteveL-MSFT replied:
Perhaps it should be called TelemetryOptoutEnvVar

diddledan replied:
See a54f7dc where I have renamed the constant as you proposed.

Commit:
Refactor TelemetryOptout constant

Rename the constant  to  to better reflect its purpose.

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -50,14 +73,18 @@ private static void SendTelemetry(string eventName, Dictionary<string,string>pay
             {
                 // if the semaphore file exists, try to send telemetry
-                if (Utils.NativeFileExists(TelemetrySemaphoreFilePath))
+                var Enabled = Utils.NativeFileExists(TelemetrySemaphoreFilePath) && !GetEnvironmentVariableAsBool(TelemetryOptoutEnvVar, false);
```
local vars should be camelCased

diddledan replied:
oops, sorry about that one :-) I think I might have accidentally copied the style from somewhere else. I'll add another commit shortly to correct it.

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -50,14 +73,18 @@ private static void SendTelemetry(string eventName, Dictionary<string,string>pay
-                if (Utils.NativeFileExists(TelemetrySemaphoreFilePath))
+                var Enabled = Utils.NativeFileExists(TelemetrySemaphoreFilePath) && !GetEnvironmentVariableAsBool(TelemetryOptoutEnvVar, false);
+
+                if ( ! Enabled )
```
for consistency (like line 76 above), remove the whitespace between `!` and `enabled`

diddledan replied:
haha, that is the WordPress coding style.. Polyglots: knowledge of many things, master of none. I'll fold that into the same commit as the camelCased `Enabled` variable.

Commit:
Fix coding style

Rename `Enabled` to `enabled` for conformance to camelCase local variables.
Remove excess whitespace in `if` constructs.

diddledan commented:
I've updated the PR again

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/Telemetry.cs:
```diff
@@ -42,6 +43,28 @@ static ApplicationInsightsTelemetry()
             TelemetryConfiguration.Active.TelemetryChannel.DeveloperMode = _developerMode;
         }
 
+        internal static bool GetEnvironmentVariableAsBool(string name, bool defaultValue) {
```
is there a reason for `internal` instead of `private`?  if you think this would be generally useful, perhaps this method should be `psutils.cs` instead of this file.  I would be ok with just making this private for now and in the future if this method is used elsewhere we can move it.

diddledan replied:
To be honest, I think the `internal` came from dotnet-cli which is where I got the inspiration for this function. I have no rationale of choosing one over the other so I am happy with whichever you prefer.

SteveL-MSFT replied:
I would suggest leaving it here and changing to `private`.

SteveL-MSFT reviewed (changes requested):
Thanks for the contribution.  Sorry about the stream of feedback, been juggling multiple things today.

Commit:
Change signature of GetEnvironmentVariableAsBool

Change visibility of GetEnvironmentVariableAsBool from `internal static` to `private static`

diddledan commented:
I've updated the PR.
